### ISSUE TITLE
Cache relations in DBT

### ISSUE THREAD

**Issue by beckjake** [fishtown-analytics/dbt#911]
## Relation Caching plan
This is a general roadmap for how we can implement a relation caching mechanism for dbt adapters. Instead of `list_relations()` going out to the database every time, we can maintain a local cache in between adapter invocations. As long as we're careful to invalidate and update the cache as appropriate, we can save a lot of time on metadata queries.

### New concepts
Relations will be cached on a per-schema basis.

For a given schema, its cache can be fully valid, partially valid, or not present at all.

A fully valid cache means that no changes have been made since the last call to list_relations except via `drop_relation`, `truncate_relation`, or `rename_relation` - so the cache is 100% up to date (see below for new methods around this).
	- note that creating a new relation will mean it isn't fully valid under this plan

A partially valid cache means that there may be more relations that aren't in the list of relations
	- so for any calls to `list_relations` or `get_relation` with an identifier that isn't present, the cache must be refreshed

When a relation in a schema is invalidated, it is removed from the cache, and the cache is marked as partially valid.

#### new type `CacheInvalid(Exception)`
Internal control-flow exception. Raised when the cache is not fully valid and the caller needs to re-populate the cache and call again.

#### new type `RelationsCache`
Stores per-schema relations caches

##### methods
- `__contains__(self, schema)`:
	- Returns True if the schema is cached.
- `clear(self)`: Clears the schema for all schemas.
	- No return value
- `clear_schema(self, schema)`: Clears the schema if it's present, otherwise does nothing.
	- No return value.
- `invalidate_schema(self, schema)`: Marks a schema as not being fully valid, if it's present. Otherwise does nothing.
	- No return value.
- `rename_relation(self, from_relation, to_relation)`: Replace the old relation with a new one, if present. Otherwise insert it (or error?).
	- No return value.
- `clear_relation(self, relation)`: Invalidate a single relation by removing it from the schema cache, if present. Otherwise does nothing (or error?).
	- No return value.
- `list_relations(self, schema)`: See SchemaCache.list_relations
	- Raises `CacheInvalid` if the schema is not in the cache at all
- `get_relation(self, schema, search)`: See SchemaCache.get_relation.
- `add_relation(self, relation)`: See SchemaCache.add_relation.
	- Creates a new partially valid SchemaCache if none exists.

#### new (internal) type `SchemaCache`
Store a single schema's relations caches. I don't think this has to be exposed anywhere outside of the `RelationsCache`.

##### attributes
- `fully_valid`: `bool` indicating that the cache is a complete representation of the relations in the schema

##### methods
- `__contains__(self, relation)`:
	- Returns True if the relation is in the cache
- `clear_relation(self, identifier)`: Invalidate a single relation by removing it from the schema cache. (TODO: ensure we don't need realationship_type!)
	- No return value.
- `add_relation(self, relation)`: Add a new relation to the schema cache.
	- No return value.
- `list_relations(self)`: Get the list of relations cached for this schema
	- Returns a list of Relations if the cache is fully valid
	- Raises `CacheInvalid` if the cache is not fully valid
- `get_relation(self, search)`: Search for the relation in the cache.
	- Returns a list of Relations that match (caller must handle duplicates).
	- Raises `CacheInvalid` if the relation is not found and the schema cache is partially valid, raise `CacheInvalid` instead of returning an empty list.
	- if the relation is found and the schema cache is partially valid, we assume there are no dupliates since we filled the cache.

#### Update `DefaultAdapter` and subclasses
- They will now call into the RelationsCache to update it in `drop_relation`, `rename_relation`
	- `drop_relation` does a cascade, so we will have to invalidate the full schema on drops
		- do we have to invalidate all schemas?
	- `rename_relation` should be fine.
- They will look at the cache for `get_relation`/`list_relations`, catch any `CacheInvalid` and update the cache appropriately
- Add an `execute_hook(cls, profile, sql, ...)` method that parses the SQL before running
	- detect 'destructive' changes
		- `drop ...`
		- `alter ...`
		- invalidate just the schema if we can tell it matches, otherwise we need to invalidate the full cache (with `cascade`, does it have to be the entire cache all the time?)
	- call execute()
- Add an `execute_model`(cls, profile, schema, identifier, relation_type, sql, ...) method
	- add the new relation (it will always be an addition because we rename existing tables)
	- call execute()

#### Update Runner.run_hooks
- Instead of calling `adapter.execute()` in `run_hooks`, call the new `execute_hook` above

#### Update macros
- Macros that create things will also need to add those relations to the cache:
	- `create_view_as`
	- `create_table_as`
	- `create_archive_table`?

#### Add new module-level fields
- `dbt/adapters/__init__.py` gets a new field, `relations_cache` that is initialized to an empty RelationCache
	- I think integration tests will have to clear this in setup/teardown

### Alternatives
Instead of raising a control-flow exception, we can change anything that would raise `CacheInvalid` to return `None` and make sure we always properly inspect return values of functions.

Instead of putting the relations cache at the global level, instantaite adapters where we currently return the types, and make everything instance methods instead of classmethods.

Instead of updating macros, we could try to introspect into the query inside adapter.execute. We don't know the schema there, or even that we are executing a model, so we'd have to extract it from the parsed query.

Instead of adding `execute_model` and changing the macros to call it, the macros could call the
We could also update the macros to call a new `execute_model` like the `execute_hook` described above, and handle caching there where we can pass along schemas and such.

We might not need to have a per-schema cache if we only ever care about the target schema, but I think it might matter for archives? If it doesn't, we can get rid of the `RelationCache` and combine it into the `SchemaCache`

**Comment by beckjake** [comment]
@drewbanin 
1. Nope, that was a mistake on my part.
2. I agree. The `relation` field in those caches is always a `Relation`, which includes the `relation_type` field. `clear_relation()` does take just the identifier, but it's remove only and only operates on the cache, so I don't think it needs the type. All it has to do is iterate over the cache members, find the entry with a matching identifier, and remove it. Unless some databases allow views and tables to have the same name, is that the concern?
3. Is it possible to ask the database what a cascade would delete? Or heck, even did delete? I do think this is possible though, I'll look into that some more.
4. I'm fine with letting hooks do whatever.

**Comment by cmcarthur** [comment]
> we should leverage the properties of the DAG where appropriate.

Agree. This may inform the structure of the `SchemaCache` / `RelationCache` object -- it may end up looking more catalog-y.

> I think that running any sort of statements that directly affect the dbt graph is an anti-pattern, and it should be ok for caching to throw its hands up in that scenario.

I agree, but it's important that when caching "throws its hands up," it does so in a way that doesn't cause radical behavior. We have, as a rule, stayed away from SQL parsing to date, but I don't know that we'll be able to continue to do so. If we need to parse SQL to do this right, let's do it.

**Comment by cmcarthur** [comment]
@beckjake I think it's important to write down the pathological case for our materializations and caching scenarios. Materializations should, as much as possible, avoid leaving the warehouse in an incomplete state. In RDBMSes, this is easy to reason about because everything is transactional -- if the cache says a table exists, but you get an error when trying to ALTER it, you can usually roll back the entire transaction. This isn't the case for the other cloud data warehouses.

If a user of dbt were to `DROP SCHEMA ... CASCADE` in the middle of their dbt run, I think that dbt would not do anything nasty. It'd just abort mid-run. I suspect that there are other types of queries that could cause real problems, but I don't know that for a fact. It could be that dbt would handle these situations well. Let's enumerate the error cases and only address the bad ones, and maybe we can forgo things like SQL parsing.

**Comment by drewbanin** [comment]
@beckjake to your third point:
> Is it possible to ask the database what a cascade would delete? Or heck, even did delete? I do think this is possible though, I'll look into that some more.

I don't think there's a general way to find this info out cross-platform. Or, if there is, it would be no better than just running the actual introspective query to see what still exists!

**Comment by beckjake** [comment]
I guess the only real "risk" is the risk of a query failing in the middle of a dbt run, since `drop ... cascade` is only going to delete stuff and then we'd be `select`ing from missing tables or renaming missing tables or something. I think that would be fine, even if you can't roll back, because you could just rebuild all your models, right? It might take a long time is all.

**Comment by drewbanin** [comment]
I wrote this like it was a computer science paper. It's probably needlessly verbose, but I wanted to define a language that we can use to talk about caching. This is info about how we can use 1) the dbt DAG and 2) the properties of modern warehouses to cache relations intelligently:

## Caching Strategy

### Overview
There are three classes of database operations that are cache-mutative:
- Create a schema or relation
- Drop a schema or relation
- Swap a relation

All “create” operations are “local”; that is: they only mutate the cache for the entity that is being created. Both the “drop” and “swap” classes operations are “nonlocal” in nature. For these nonlocal operations, other entities in the cache may need to be mutated as a result of the nonlocal operations. The exact cache mutation logic is dependent on 1) the type of warehouse in use and 2) the type of relations present in the database.

### Creative operations
In the `create` case, a simple addition to the cache is required. No further cache mutation is required.

### Destructive operations

#### Dropping a relation
Database relations can be “structurally dependent” on other relations. This structural dependency is a function of one relation selecting from another relation. The term “structurally dependent” is used to denote a tighter dependency than the typical ”logical dependency” inherent in dbt.

##### Logical Dependency
Tables can select from other relations. When this happens, the table and the relation(s) that it selects from are logically dependent. In this scenario, the selecting table maintains its own schema and data, so there is no structural dependency present. Instead, the relationship between a table and it’s parent is a logical dependency.

##### Structural Dependency
Contrastingly, views can also select from other relations. A view does not maintain its own schema or data — rather, the schema and data that comprise a view are a function of 1) the view query definition and 2) the relations that it selects from. Because the definition of a view is a function of the relations that it selects from, it is said to be “structurally” dependent on its parents. More formally: structurally dependent relationships exist where the definition for one relation is inherently tied to the existence of a separate relation. Crucially, this structural dependency is recursive in nature.

Consider the following example:

```
+--------------+               +------------------+          +-------------------+
|              +--------------->                  |          |                   |
| Table A      |               |   View B         +---------->    View C         |
+--------------+               +------------------+          +-------------------+
```

Here, Table A is (by definition) not structurally dependent on any relations.
View B is structurally dependent on Table A.
View C is structurally dependent on Table B, and through recursive dependence, it is also structurally dependent on Table A.

##### Adapter-specific behavior

Both BigQuery and Snowflake have implemented “views” in a late-binding fashion. Views on these databases are logically dependent, but not structurally dependent.

In Redshift and Postgres, views are structurally dependent. Interestingly, Redshift supports Late Binding Views which are not structurally dependent.

When a relation is `drop…cascaded`, all of the structurally dependent descendants of the relation should also be marked as deleted in the cache.


#### Dropping a schema
For cache purposes, dropping a schema is equivalent to dropping all of the relations in the schema, and then dropping the schema itself. When a schema is dropped, all of the relations that structurally depend on relations inside of that schema should also be dropped.

**Comment by drewbanin** [comment]
I didn't actually touch on the renaming piece... need to think about that one a little bit more

**Comment by beckjake** [comment]
Renaming is just a drop + a create from a cache perspective, right?

**Comment by drewbanin** [comment]
@beckjake that's right. I wanted to investigate if there's any sort of difference in how postgres/redshift handles the drop-and-swap if we do it inside of a transaction. I don't think it makes any difference, but I wanted to confirm.

**Comment by drewbanin** [comment]
After chatting with Jake:

The dbt graph is an approximation of the the structural dependency graph inherent in the database. In situations where users use `{{ ref() }}` outside of a select statement (eg. in a comment (eg. to hoist refs)), then the dbt graph and the database dependency graph might not be exactly equivalent.

Instead, dbt should built a _separate_ graph of structural dependencies by querying the database at the beginning of the run.

**Comment by beckjake** [comment]
It's worth noting that we only have to build the graph for postgres and redshift, and for redshift we can completely ignore late-binding views (since all we really care about is "will my `drop ... cascade` or `alter table ... rename` change anything?").
